tools/sof-perf-analyzer.py

```diff
@@ -0,0 +1,253 @@
+    timestamp: float
+    samples: int
+    period: int
+    cpu_avg: int
+    cpu_peak: int
+
+@dataclass(frozen=True)
+class CompPerfStats:
+    '''The dataclass for SOF component statistics'''
+    avg_cpu_avg: float   # average of cpu average
+    avg_cpu_peak: float  # average of cpu peak
+    peak_cpu_avg: float  # peak of cpu average
+    peak_cpu_peak: float # peak of cpu peak
+
+# Currently, we don't have test case to test low-power mode, in which
+# the DSP may not operate in the maximum clock frequency. Let't hardcode
+# this value to the maximum clock frequency for now.
+DSP_CLK = 400000000
+# The DSP timer clock frequency is fixed despite of DSP_CLK
+DSP_TIMER = 38400000
+UINT32_MAX = 4294967295
+AUDIO_PERIOD = 1000
+
+TICK_TO_MCPS = DSP_CLK / DSP_TIMER / AUDIO_PERIOD
+
+# Once DSP timer count to UINT32_MAX, it wrapped to zero, which causes
+# the timestamp wrap to zero, too. This ts_shift is used to correct
+# the timestamp value on wrap. UINT32_MAX / DSP_TIMER is added to it at
+# every wrap.
+#
+# pylint: disable=C0103
+ts_shift = 0
+# pylint: disable=C0103
+args = None
+
+# The dictionary for performance information
+perf_info: Dict[Comp, List[CompPerfInfo]] = {}
+# The dictionary for performace statistics
+perf_stats: Dict[Comp, CompPerfStats] = {}
+# Component ID to name map acquired from dmesg
+comp_name_map: Dict[Comp, str] = {}
+
+TraceItemGenerator = Generator[TraceItem, None, None]
+
+def collect_perf_info(trace_item: TraceItem):
+    '''Collect trace information in mtrace to a Comp->List[CompPerfInfo] mapping'''
+    msg = trace_item.msg.split()
+    ppl_id = msg[0].split(':')[1]
+    perf_val = int(msg[5]), int(msg[7]), int(msg[10]), int(msg[12])
+    perf_info.setdefault(Comp(ppl_id, msg[1]),[]).append(
+        CompPerfInfo(trace_item.timestamp, *perf_val)
+        )
+
+def dispatch_trace_item(trace_item: TraceItem):
+    '''Dispatch trace item to cosponding analyzing function'''
+    if trace_item.func == 'comp_copy':
+        collect_perf_info(trace_item)
+
+def get_first_valid_trace(trace_item_gen: TraceItemGenerator) -> Optional[TraceItem]:
+    '''The current sof-test test case may collect some traces belongs to previous
+    test case due to mtrace is configured in deferred mode. This function consumes
+    those invalid traces from the generator, and return the first valid trace item.
+    '''
+    try:
+        while item := next(trace_item_gen):
+            if int(item.timestamp) == 0:
+                return item
+    except StopIteration:
+        return None
+    return None
+
+def make_trace_item(fileio: TextIO) -> TraceItemGenerator:
+    '''Convert a line of trace in string form into TraceItem object, for example:
+    '[0.349880] <inf> pipe: pipeline_connect: comp:0 0x40000 connect buffer 0 as sink' ->
+    TraceItem(timestamp=0.349880, level='inf', context='pipe', func='pipeline_connect',
+              msg='comp:0 0x40000 connect buffer 0 as sink')
```

Comment:
So the entire purpose of this script is to parse performance information from the logs but there's no obvious performance information in the log example...?

Also, please have another (or the same) example like this at the very top.